tests/ui/cognitive_complexity.rs

```diff
@@ -369,3 +359,32 @@ fn early_ret() -> i32 {
         _ => return 6,
     }
 }
+
+#[clippy::cognitive_complexity = "0"]
+fn osscilating_logical_chain_1() -> bool {
+    true && false || true && false
+}
+
+// This tests that the only thing that matters 
+// is the change in operator
+#[clippy::cognitive_complexity = "0"]
+fn osscilating_logical_chain_2() -> bool {
+    true && false && true && false || true && false && true && false
+}
+
+#[clippy::cognitive_complexity = "0"]
+fn osscilating_logical_chain_3() -> bool {
+    (true && false) || (true && false)
+}
+
+#[clippy::cognitive_complexity = "1"]
+fn nested_functions_are_counted(n: usize) -> usize {
+    fn lucas_number(n: usize) -> usize {
+        match n {
+            0 => 2,
+            1 => 1,
+            _ => lucas_number(n - 1) + lucas_number(n - 2),
+        }
+    }
+    lucas_number(n)
+}
```

Comment:
It would be good to have a test that this doens't trigger in external macros. To do that you'd add a new auxiliary file that contains a proc macro which contains code triggering the lint. Similar to [`tests/ui/auxiliary/proc_macro_derive.rs`](https://github.com/phansch/rust-clippy/blob/60a1759b5f2af3b15ac32b5984001e3986c4b9b1/tests/ui/auxiliary/proc_macro_derive.rs) but probably better as a separate file.

Reply:
Not only that! Also a test with a local macro :D since this lint is supposed to be run before macro expansion ^^

Reply:
I've made a test for local macros, that expands to this function and a call to it. It's in `ui/cognitive_complexity/no_ops.rs`

I will make another one for non-local macros as soon as I know how!

Reply:
Thing is, while writing that response I realized that you were talking about procedural macros. Where should I learn how to make one? :D